### PR TITLE
Add support for C_DeriveKey PKCS#11 function

### PR DESCRIPTION
**Pull request by balsingh1991**
Add support for CK_KEY_DERIVATION_STRING_DATA struct for use as mechanism parameters
Add support for DUKPT key derivation
Add support for ECIES encrypt/decrypt operations
Note: For DUKPT and ECIES operations, underlying cryto implementation must support it otherwise CKR_MECHANISM_INVALID error is encountered.

### CODE REVIEW

**Comment by LudovicRousseau** [comment]
It looks nice.

Can yo provide a Unitary Test as the ones in the `test/` directory?
The idea is to:
- test the wrapper is working as expected
- check for regression in case of future code change
- "document" how to use the function

**Comment by balsingh1991** [comment]
I had created few unit tests. Unfortunately, for my use case mechanisms involved are not standard, thus, softhsm will not support it.
Trying to include derive test for CKM_ECDH1_DERIVE, but getting CKR_MECHANISM_PARAM_INVALID from softhsm. I'll add another commit when I resolve it.
Thanks!

**Comment by LudovicRousseau** [comment]
Allo @balsingh1991, any progress on the use of SoftHSM with your patches?

**Comment by balsingh1991** [comment]
Hi,

My apologies for leaving this stray for long. I will check and report
sometime next week.

Thank you for bringing this to my attention!


- Baljinder Singh

File: src/pykcs11.i [hunk 1]
```diff
@@ -1023,6 +1082,9 @@ typedef unsigned long CK_RV;
 #define CKM_HKDF_DATA                  0x0000402b
 #define CKM_HKDF_KEY_GEN               0x0000402c
 #define CKM_VENDOR_DEFINED             0x80000000UL
+#define CKM_DES2_DUKPT_MAC             (CKM_VENDOR_DEFINED + 0x612)
```
Comment (LudovicRousseau) [inline src/pykcs11.i:1085]: I can't include vendor specific mechanisms.

Are these mechanisms planned for a future version of PKCS#11 standard?

**Review by LudovicRousseau — CHANGES_REQUESTED** [review]
It would be perfect if you could:
- rebase your patches above master
- split the patches with only one new feature per patch

**Pull request closed without merging**